examples/crd_derive_schema.rs

```diff
@@ -0,0 +1,244 @@
+use anyhow::{anyhow, Result};
+use futures::{StreamExt, TryStreamExt};
+use k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition;
+use kube::CustomResource;
+use kube::{
+    api::{Api, DeleteParams, ListParams, PostParams, Resource, WatchEvent},
+    Client,
+};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+// This example shows how the generated schema affects defaulting and validation.
+// The integration test `crd_schema_test` in `kube-derive` contains the full CRD JSON generated from this struct.
+//
+// References:
+// - https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting
+// - https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting-and-nullable
+
+#[derive(CustomResource, Serialize, Deserialize, Default, Debug, PartialEq, Clone, JsonSchema)]
+#[kube(
+    group = "clux.dev",
+    version = "v1",
+    kind = "Foo",
+    namespaced,
+    derive = "PartialEq",
+    derive = "Default"
+)]
+#[kube(apiextensions = "v1")]
+pub struct FooSpec {
+    // Non-nullable without default is required.
+    //
+    // There shouldn't be any ambiguity here.
+    non_nullable: String,
+
+    // Non-nullable with default value.
+    //
+    // Serializing will work as expected because the field cannot be `None`.
+    //
+    // When deserializing a response from the server, the field should always be a string because
+    // the field is non-nullable and the server sets the value to the default specified in the schema.
+    //
+    // When deserializing some input, the default value will be set if missing.
+    // However, if `null` is specified, `serde` will panic.
+    // The server prunes `null` for non-nullable field since 1.20 and the default is applied.
+    // To match the server's behavior exactly, we can use a custom deserializer.
+    #[serde(default = "default_value")]
+    non_nullable_with_default: String,
+
+    // Nullable without default, skipping None.
+    //
+    // By skipping to serialize, the field won't be present in the object.
+    // If serialized as `null` (next field), the object will have the field set to `null`.
+    //
+    // Deserializing works as expected either way. `None` if it's missing or `null`.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    nullable_skipped: Option<String>,
```

Comment:
The documentation here is really on point.
One note perhaps worth adding is that `nullable_skipped` will not let you override a set value without using a hacky `patch` as serde will override any explicit `None`s.

Reply:
Although, studying the example further shows this is covered well by the behaviour, and it's not like this is API documentation.